### ISSUE TITLE
Not enough memory, cannot connect to backend

### ISSUE THREAD

**Issue by BlackYoup** [sozu-proxy/sozu#416]
I just had this issue. I think it got triggered when I shutdown one of my backends and some clients were still doing some requests.

full error:
```
Mar 26 21:04:12 exherbo-dedicated sozu[10784]: 2018-03-26T19:04:09Z 7001458036798255 10799 WRK-00 ERROR        error connecting to backend, trying again
Mar 26 21:04:12 exherbo-dedicated sozu[10784]: 2018-03-26T19:04:09Z 7001458036807565 10799 WRK-00 ERROR        not enough memory, cannot connect to backend
```

There may be a leak somewhere. Also, this was looping indefinitely, making the worker unresponsive to any other events.

**Comment by BlackYoup** [comment]
I can't find a way to reproduce locally but it gets triggered quite easily on my test server. Before the leak happens, the `error connecting to backend, trying again` loops for a while. There may be a problem in the retry policy which then triggers the leak

**Comment by BlackYoup** [comment]
As a side node, I have this issue with TLS requests, I didn't test with plain HTTP requests.

**Comment by Geal** [comment]
this might be fixed by f15dc08, please test

**Comment by BlackYoup** [comment]
It does not leak anymore but there is still a problem with the retry policy. I'll open another issue
